File: tools/editor-server.mjs
#!/usr/bin/env node
/* =========================================================================
   Portfolio editor server — LOCAL ONLY.
   Serves the static site from the repo root and injects a WYSIWYG editing
   layer that writes edits back to the real files. The editor assets are
   injected at request time, so the files on disk (and the GitHub Pages
   deploy) never reference the editor. Run with:  node tools/editor-server.mjs
   Zero dependencies — Node built-ins only.
   ========================================================================= */

import http from 'node:http';
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import { existsSync } from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');      // repo root (the site)
const EDITOR_DIR = path.join(__dirname, 'editor'); // tools/editor
const PORT = process.env.PORT ? Number(process.env.PORT) : 4321;

const MIME = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'text/javascript; charset=utf-8',
  '.mjs': 'text/javascript; charset=utf-8',
  '.json': 'application/json; charset=utf-8',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.gif': 'image/gif',
  '.svg': 'image/svg+xml',
  '.webp': 'image/webp',
  '.pdf': 'application/pdf',
  '.mp4': 'video/mp4',
  '.woff': 'font/woff',
  '.woff2': 'font/woff2',
  '.ico': 'image/x-icon'
};

// --- helpers --------------------------------------------------------------

function send(res, status, body, headers = {}) {
  res.writeHead(status, { 'Cache-Control': 'no-store', ...headers });
  res.end(body);
}

function sendJson(res, status, obj) {
  send(res, status, JSON.stringify(obj), { 'Content-Type': 'application/json; charset=utf-8' });
}

function readBody(req) {
  return new Promise((resolve, reject) => {
    const chunks = [];
    req.on('data', (c) => chunks.push(c));
    req.on('end', () => resolve(Buffer.concat(chunks).toString('utf8')));
    req.on('error', reject);
  });
}

// Resolve a request path to an absolute path that MUST stay inside `base`.
function safeResolve(base, relPath) {
  const clean = decodeURIComponent(relPath).split('?')[0].replace(/^\/+/, '');
  const abs = path.resolve(base, clean);
  if (abs !== base && !abs.startsWith(base + path.sep)) return null;
  return abs;
}

// Inject the editor stylesheet, the __EDIT__ flag, and the editor module.
function injectEditor(html) {
  const head =
    '<link rel="stylesheet" href="/__editor__/editor.css">\n' +
    '<script>window.__EDIT__=true;</script>\n';
  const tail = '<script type="module" src="/__editor__/editor.js"></script>\n';
  let out = html;
  out = out.includes('</head>') ? out.replace('</head>', head + '</head>') : head + out;
  out = out.includes('</body>') ? out.replace('</body>', tail + '</body>') : out + tail;
  return out;
}

/* Replace the inner HTML of the element carrying data-edit-id="<editId>".
   Walks balanced same-tag open/close tokens, so it is safe as long as an
   editable element does not nest another element of its own tag name (none
   of ours do). Returns the new file string. */
function replaceInner(html, editId, newInner) {
  const attr = 'data-edit-id="' + editId + '"';
  const attrIdx = html.indexOf(attr);
  if (attrIdx === -1) throw new Error('data-edit-id not found: ' + editId);

  const tagStart = html.lastIndexOf('<', attrIdx);
  const nameMatch = /^<([a-zA-Z][a-zA-Z0-9-]*)/.exec(html.slice(tagStart));
  if (!nameMatch) throw new Error('could not find opening tag for: ' + editId);
  const tag = nameMatch[1];

  const openEnd = html.indexOf('>', attrIdx);
  if (openEnd === -1) throw new Error('unterminated opening tag for: ' + editId);
  const innerStart = openEnd + 1;

  const tokenRe = new RegExp('<' + tag + '(?:[\\s>/])|</' + tag + '>', 'gi');
  tokenRe.lastIndex = innerStart;
  let depth = 1;
  let innerEnd = -1;
  let m;
  while ((m = tokenRe.exec(html))) {
    if (m[0].slice(0, 2) === '</') {
      depth--;
      if (depth === 0) { innerEnd = m.index; break; }
    } else {
      depth++;
    }
  }
  if (innerEnd === -1) throw new Error('no matching </' + tag + '> for: ' + editId);
  return html.slice(0, innerStart) + newInner + html.slice(innerEnd);
}

// --- API handlers ---------------------------------------------------------

async function handleSaveData(req, res) {
  const body = await readBody(req);
  let data;
  try { data = JSON.parse(body); } catch { return sendJson(res, 400, { error: 'invalid JSON' }); }
  if (!data || !Array.isArray(data.projects)) {
    return sendJson(res, 400, { error: 'expected { projects: [...] }' });
  }
  const file = path.join(ROOT, 'data', 'projects.json');
  await writeFile(file, JSON.stringify(data, null, 2) + '\n', 'utf8');
  sendJson(res, 200, { ok: true, count: data.projects.length });
}

async function handleSaveText(req, res) {
  const body = await readBody(req);
  let payload;
  try { payload = JSON.parse(body); } catch { return sendJson(res, 400, { error: 'invalid JSON' }); }
  const { file, editId, html } = payload || {};
  if (!file || !editId || typeof html !== 'string') {
    return sendJson(res, 400, { error: 'expected { file, editId, html }' });
  }
  if (!/^[\w-]+\.html$/.test(file)) {
    return sendJson(res, 400, { error: 'file must be a top-level .html file' });
  }
  const abs = path.join(ROOT, file);
  if (!existsSync(abs)) return sendJson(res, 404, { error: 'file not found: ' + file });
  try {
    const current = await readFile(abs, 'utf8');
    const updated = replaceInner(current, editId, html);
    await writeFile(abs, updated, 'utf8');
    sendJson(res, 200, { ok: true });
  } catch (err) {
    sendJson(res, 500, { error: String(err.message || err) });
  }
}

async function handleUpload(req, res) {
  const body = await readBody(req);
  let payload;
  try { payload = JSON.parse(body); } catch { return sendJson(res, 400, { error: 'invalid JSON' }); }
  let { name, dataBase64, dir } = payload || {};
  if (!name || !dataBase64) return sendJson(res, 400, { error: 'expected { name, dataBase64, dir }' });

  // Sanitize: keep a flat, safe filename and a safe sub-directory under assets/.
  name = String(name).replace(/[^\w.\-]+/g, '_').replace(/^\.+/, '');
  dir = String(dir || 'uploads').replace(/[^\w/\-]+/g, '_').replace(/\.\.+/g, '_').replace(/^\/+|\/+$/g, '');
  const targetDir = path.join(ROOT, 'assets', dir);
  if (!targetDir.startsWith(path.join(ROOT, 'assets'))) {
    return sendJson(res, 400, { error: 'invalid dir' });
  }

  // strip a possible data: URI prefix
  const comma = dataBase64.indexOf(',');
  const raw = dataBase64.startsWith('data:') && comma !== -1 ? dataBase64.slice(comma + 1) : dataBase64;

  try {
    await mkdir(targetDir, { recursive: true });
    // avoid clobbering: if it exists, suffix with a counter
    let finalName = name;
    let n = 1;
    while (existsSync(path.join(targetDir, finalName))) {
      const dot = name.lastIndexOf('.');
      finalName = dot === -1 ? name + '-' + n : name.slice(0, dot) + '-' + n + name.slice(dot);
      n++;
    }
    await writeFile(path.join(targetDir, finalName), Buffer.from(raw, 'base64'));
    // Return a web path, URL-encoded per segment (matches projects.json style).
    const webPath = 'assets/' + dir.split('/').map(encodeURIComponent).join('/') + '/' + encodeURIComponent(finalName);
    sendJson(res, 200, { ok: true, path: webPath });
  } catch (err) {
    sendJson(res, 500, { error: String(err.message || err) });
  }
}

// --- static serving -------------------------------------------------------

async function serveStatic(req, res, base, urlPath, { inject = false } = {}) {
  let abs = safeResolve(base, urlPath);
  if (!abs) return send(res, 403, 'Forbidden');
  if (urlPath === '/' || urlPath === '') abs = path.join(base, 'index.html');
  if (existsSync(abs) && (await import('node:fs')).statSync(abs).isDirectory()) {
    abs = path.join(abs, 'index.html');
  }
  if (!existsSync(abs)) return send(res, 404, 'Not found: ' + urlPath);

  const ext = path.extname(abs).toLowerCase();
  const type = MIME[ext] || 'application/octet-stream';
  try {
    if (ext === '.html' && inject) {
      const html = await readFile(abs, 'utf8');
      return send(res, 200, injectEditor(html), { 'Content-Type': type });
    }
    const buf = await readFile(abs);
    send(res, 200, buf, { 'Content-Type': type });
  } catch (err) {
    send(res, 500, 'Read error: ' + err.message);
  }
}

// --- router ---------------------------------------------------------------

const server = http.createServer(async (req, res) => {
  const urlPath = req.url.split('?')[0];

  // editor assets
  if (urlPath.startsWith('/__editor__/')) {
    if (req.method === 'POST') {
      if (urlPath === '/__editor__/save-data') return handleSaveData(req, res);
      if (urlPath === '/__editor__/save-text') return handleSaveText(req, res);
      if (urlPath === '/__editor__/upload-image') return handleUpload(req, res);
      return sendJson(res, 404, { error: 'unknown endpoint' });
    }
    const assetPath = urlPath.replace('/__editor__/', '/');
    return serveStatic(req, res, EDITOR_DIR, assetPath, { inject: false });
  }

  if (req.method !== 'GET') return send(res, 405, 'Method not allowed');
  return serveStatic(req, res, ROOT, urlPath, { inject: true });
});

server.listen(PORT, () => {
  console.log('\n  Portfolio editor running:  \x1b[1mhttp://localhost:' + PORT + '/\x1b[0m');
  console.log('  Serving:                   ' + ROOT);
  console.log('  Edits save to real files. The live site never loads the editor.');
  console.log('  Press Ctrl+C to stop.\n');
});

File: tools/editor_server.py
#!/usr/bin/env python3
"""
Portfolio editor server -- LOCAL ONLY.

Serves the static site from the repo root and injects a WYSIWYG editing layer
that writes edits back to the real files. The editor assets are injected at
request time, so the files on disk (and the GitHub Pages deploy) never
reference the editor.

Run with no install -- macOS ships Python 3:

    python3 tools/editor_server.py

Standard library only.
"""

import base64
import json
import mimetypes
import os
import re
import sys
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer

HERE = os.path.dirname(os.path.abspath(__file__))
ROOT = os.path.abspath(os.path.join(HERE, os.pardir))   # repo root (the site)
EDITOR_DIR = os.path.join(HERE, "editor")               # tools/editor
PORT = int(os.environ.get("PORT", "4321"))

EXTRA_MIME = {
    ".js": "text/javascript; charset=utf-8",
    ".mjs": "text/javascript; charset=utf-8",
    ".css": "text/css; charset=utf-8",
    ".json": "application/json; charset=utf-8",
    ".svg": "image/svg+xml",
    ".webp": "image/webp",
    ".woff2": "font/woff2",
}


def guess_type(path):
    ext = os.path.splitext(path)[1].lower()
    if ext in EXTRA_MIME:
        return EXTRA_MIME[ext]
    if ext in (".html", ".htm"):
        return "text/html; charset=utf-8"
    t = mimetypes.guess_type(path)[0]
    return t or "application/octet-stream"


def inject_editor(html):
    """Inject the editor stylesheet, the __EDIT__ flag, and the editor module."""
    head = (
        '<link rel="stylesheet" href="/__editor__/editor.css">\n'
        "<script>window.__EDIT__=true;</script>\n"
    )
    tail = '<script type="module" src="/__editor__/editor.js"></script>\n'
    if "</head>" in html:
        html = html.replace("</head>", head + "</head>", 1)
    else:
        html = head + html
    if "</body>" in html:
        html = html.replace("</body>", tail + "</body>", 1)
    else:
        html = html + tail
    return html


def replace_inner(html, edit_id, new_inner):
    """Replace the inner HTML of the element carrying data-edit-id="<edit_id>".

    Walks balanced same-tag open/close tokens, so it is safe as long as an
    editable element does not nest another element of its own tag name (none
    of ours do).
    """
    attr = 'data-edit-id="%s"' % edit_id
    i = html.find(attr)
    if i == -1:
        raise ValueError("data-edit-id not found: " + edit_id)

    tag_start = html.rfind("<", 0, i)
    m = re.match(r"<([a-zA-Z][\w-]*)", html[tag_start:])
    if not m:
        raise ValueError("could not find opening tag for: " + edit_id)
    tag = m.group(1)

    open_end = html.find(">", i)
    if open_end == -1:
        raise ValueError("unterminated opening tag for: " + edit_id)
    inner_start = open_end + 1

    token_re = re.compile(r"<%s(?:[\s>/])|</%s>" % (re.escape(tag), re.escape(tag)), re.I)
    depth = 1
    inner_end = -1
    for mt in token_re.finditer(html, inner_start):
        if mt.group(0)[:2] == "</":
            depth -= 1
            if depth == 0:
                inner_end = mt.start()
                break
        else:
            depth += 1
    if inner_end == -1:
        raise ValueError("no matching </%s> for: %s" % (tag, edit_id))
    return html[:inner_start] + new_inner + html[inner_end:]


def safe_join(base, rel):
    """Resolve rel under base, refusing anything that escapes base."""
    rel = rel.split("?")[0].lstrip("/")
    abs_path = os.path.abspath(os.path.join(base, rel))
    if abs_path != base and not abs_path.startswith(base + os.sep):
        return None
    return abs_path


class Handler(BaseHTTPRequestHandler):
    server_version = "PortfolioEditor/1.0"

    # ---- helpers --------------------------------------------------------
    def _send(self, status, body=b"", ctype="text/plain; charset=utf-8", extra=None):
        if isinstance(body, str):
            body = body.encode("utf-8")
        self.send_response(status)
        self.send_header("Content-Type", ctype)
        self.send_header("Content-Length", str(len(body)))
        self.send_header("Cache-Control", "no-store")
        for k, v in (extra or {}).items():
            self.send_header(k, v)
        self.end_headers()
        if self.command != "HEAD":
            self.wfile.write(body)

    def _send_json(self, status, obj):
        self._send(status, json.dumps(obj), "application/json; charset=utf-8")

    def _read_json(self):
        length = int(self.headers.get("Content-Length", "0"))
        raw = self.rfile.read(length) if length else b""
        return json.loads(raw.decode("utf-8"))

    def log_message(self, fmt, *args):  # quieter logging
        sys.stderr.write("  %s\n" % (fmt % args))

    # ---- routing --------------------------------------------------------
    def do_GET(self):
        path = self.path.split("?")[0]
        if path.startswith("/__editor__/"):
            asset = path[len("/__editor__/"):]
            return self.serve_static(EDITOR_DIR, asset, inject=False)
        return self.serve_static(ROOT, path, inject=True)

    def do_HEAD(self):
        self.do_GET()

    def do_POST(self):
        path = self.path.split("?")[0]
        try:
            if path == "/__editor__/save-data":
                return self.handle_save_data()
            if path == "/__editor__/save-text":
                return self.handle_save_text()
            if path == "/__editor__/upload-image":
                return self.handle_upload()
        except Exception as e:  # noqa: BLE001
            return self._send_json(500, {"error": str(e)})
        return self._send_json(404, {"error": "unknown endpoint"})

    # ---- static ---------------------------------------------------------
    def serve_static(self, base, rel, inject):
        abs_path = safe_join(base, rel)
        if abs_path is None:
            return self._send(403, "Forbidden")
        if rel in ("", "/") or os.path.isdir(abs_path):
            abs_path = os.path.join(abs_path, "index.html")
        if not os.path.isfile(abs_path):
            return self._send(404, "Not found: " + rel)
        ctype = guess_type(abs_path)
        try:
            if inject and abs_path.lower().endswith((".html", ".htm")):
                with open(abs_path, "r", encoding="utf-8") as f:
                    return self._send(200, inject_editor(f.read()), ctype)
            with open(abs_path, "rb") as f:
                return self._send(200, f.read(), ctype)
        except OSError as e:
            return self._send(500, "Read error: " + str(e))

    # ---- API ------------------------------------------------------------
    def handle_save_data(self):
        data = self._read_json()
        if not isinstance(data, dict) or not isinstance(data.get("projects"), list):
            return self._send_json(400, {"error": "expected { projects: [...] }"})
        target = os.path.join(ROOT, "data", "projects.json")
        with open(target, "w", encoding="utf-8") as f:
            f.write(json.dumps(data, indent=2, ensure_ascii=False) + "\n")
        return self._send_json(200, {"ok": True, "count": len(data["projects"])})

    def handle_save_text(self):
        payload = self._read_json()
        fname = payload.get("file")
        edit_id = payload.get("editId")
        html = payload.get("html")
        if not fname or not edit_id or not isinstance(html, str):
            return self._send_json(400, {"error": "expected { file, editId, html }"})
        if not re.match(r"^[\w-]+\.html$", fname):
            return self._send_json(400, {"error": "file must be a top-level .html file"})
        abs_path = os.path.join(ROOT, fname)
        if not os.path.isfile(abs_path):
            return self._send_json(404, {"error": "file not found: " + fname})
        with open(abs_path, "r", encoding="utf-8") as f:
            current = f.read()
        updated = replace_inner(current, edit_id, html)
        with open(abs_path, "w", encoding="utf-8") as f:
            f.write(updated)
        return self._send_json(200, {"ok": True})

    def handle_upload(self):
        payload = self._read_json()
        name = payload.get("name")
        data_b64 = payload.get("dataBase64")
        sub = payload.get("dir") or "uploads"
        if not name or not data_b64:
            return self._send_json(400, {"error": "expected { name, dataBase64, dir }"})

        name = re.sub(r"[^\w.\-]+", "_", str(name)).lstrip(".")
        sub = re.sub(r"\.\.+", "_", re.sub(r"[^\w/\-]+", "_", str(sub))).strip("/")
        target_dir = os.path.abspath(os.path.join(ROOT, "assets", sub))
        assets_root = os.path.join(ROOT, "assets")
        if not target_dir.startswith(assets_root):
            return self._send_json(400, {"error": "invalid dir"})

        if data_b64.startswith("data:") and "," in data_b64:
            data_b64 = data_b64.split(",", 1)[1]
        raw = base64.b64decode(data_b64)

        os.makedirs(target_dir, exist_ok=True)
        final = name
        n = 1
        while os.path.exists(os.path.join(target_dir, final)):
            stem, ext = os.path.splitext(name)
            final = "%s-%d%s" % (stem, n, ext)
            n += 1
        with open(os.path.join(target_dir, final), "wb") as f:
            f.write(raw)

        from urllib.parse import quote
        web = "assets/" + "/".join(quote(p) for p in sub.split("/")) + "/" + quote(final)
        return self._send_json(200, {"ok": True, "path": web})


def main():
    os.chdir(ROOT)
    httpd = ThreadingHTTPServer(("127.0.0.1", PORT), Handler)
    url = "http://localhost:%d/" % PORT
    print("\n  Portfolio editor running:  \033[1m%s\033[0m" % url)
    print("  Serving:                   %s" % ROOT)
    print("  Edits save to real files. The live site never loads the editor.")
    print("  Press Ctrl+C to stop.\n")
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n  Stopped.")
        httpd.server_close()


if __name__ == "__main__":
    main()
